File: frontend/js/translate.js
/**
 * Translation module for TranslateGemma integration
 */

const translate = (function() {
    // Supported languages (sorted alphabetically by name)
    const LANGUAGES = [
        { code: 'af', name: 'Afrikaans' },
        { code: 'sq', name: 'Albanian' },
        { code: 'am', name: 'Amharic' },
        { code: 'ar', name: 'Arabic' },
        { code: 'hy', name: 'Armenian' },
        { code: 'az', name: 'Azerbaijani' },
        { code: 'eu', name: 'Basque' },
        { code: 'be', name: 'Belarusian' },
        { code: 'bn', name: 'Bengali' },
        { code: 'bs', name: 'Bosnian' },
        { code: 'bg', name: 'Bulgarian' },
        { code: 'ca', name: 'Catalan' },
        { code: 'zh', name: 'Chinese' },
        { code: 'hr', name: 'Croatian' },
        { code: 'cs', name: 'Czech' },
        { code: 'da', name: 'Danish' },
        { code: 'nl', name: 'Dutch' },
        { code: 'en', name: 'English' },
        { code: 'et', name: 'Estonian' },
        { code: 'fi', name: 'Finnish' },
        { code: 'fr', name: 'French' },
        { code: 'gl', name: 'Galician' },
        { code: 'ka', name: 'Georgian' },
        { code: 'de', name: 'German' },
        { code: 'el', name: 'Greek' },
        { code: 'gu', name: 'Gujarati' },
        { code: 'he', name: 'Hebrew' },
        { code: 'hi', name: 'Hindi' },
        { code: 'hu', name: 'Hungarian' },
        { code: 'is', name: 'Icelandic' },
        { code: 'id', name: 'Indonesian' },
        { code: 'ga', name: 'Irish' },
        { code: 'it', name: 'Italian' },
        { code: 'ja', name: 'Japanese' },
        { code: 'kn', name: 'Kannada' },
        { code: 'kk', name: 'Kazakh' },
        { code: 'ko', name: 'Korean' },
        { code: 'lv', name: 'Latvian' },
        { code: 'lt', name: 'Lithuanian' },
        { code: 'mk', name: 'Macedonian' },
        { code: 'ms', name: 'Malay' },
        { code: 'ml', name: 'Malayalam' },
        { code: 'mt', name: 'Maltese' },
        { code: 'mr', name: 'Marathi' },
        { code: 'mn', name: 'Mongolian' },
        { code: 'ne', name: 'Nepali' },
        { code: 'no', name: 'Norwegian' },
        { code: 'fa', name: 'Persian' },
        { code: 'pl', name: 'Polish' },
        { code: 'pt', name: 'Portuguese' },
        { code: 'pa', name: 'Punjabi' },
        { code: 'ro', name: 'Romanian' },
        { code: 'ru', name: 'Russian' },
        { code: 'sr', name: 'Serbian' },
        { code: 'si', name: 'Sinhala' },
        { code: 'sk', name: 'Slovak' },
        { code: 'sl', name: 'Slovenian' },
        { code: 'es', name: 'Spanish' },
        { code: 'sw', name: 'Swahili' },
        { code: 'sv', name: 'Swedish' },
        { code: 'ta', name: 'Tamil' },
        { code: 'te', name: 'Telugu' },
        { code: 'th', name: 'Thai' },
        { code: 'tr', name: 'Turkish' },
        { code: 'uk', name: 'Ukrainian' },
        { code: 'ur', name: 'Urdu' },
        { code: 'uz', name: 'Uzbek' },
        { code: 'vi', name: 'Vietnamese' },
        { code: 'cy', name: 'Welsh' },
        { code: 'zu', name: 'Zulu' }
    ];

    // LocalStorage keys
    const STORAGE_KEYS = {
        sourceLang: 'ulfweb_source_lang',
        targetLang: 'ulfweb_target_lang'
    };

    // DOM elements
    let sourceLanguageSelect;
    let targetLanguageSelect;
    let sourceTextArea;
    let targetTextArea;
    let translateBtn;
    let stopBtn;
    let swapBtn;

    // State
    let abortController = null;

    /**
     * Initialize the translation module
     */
    function init() {
        // Get DOM elements
        sourceLanguageSelect = document.getElementById('source-language');
        targetLanguageSelect = document.getElementById('target-language');
        sourceTextArea = document.getElementById('source-text');
        targetTextArea = document.getElementById('target-text');
        translateBtn = document.getElementById('translate-btn');
        stopBtn = document.getElementById('translate-stop-btn');
        swapBtn = document.getElementById('swap-languages');

        // Populate language dropdowns
        populateLanguageDropdowns();

        // Load saved language preferences
        loadSavedLanguages();

        // Setup event listeners
        setupEventListeners();

        console.log('Translation module initialized');
    }

    /**
     * Populate language dropdown selects
     */
    function populateLanguageDropdowns() {
        LANGUAGES.forEach(lang => {
            const sourceOption = document.createElement('option');
            sourceOption.value = lang.code;
            sourceOption.textContent = lang.name;
            sourceLanguageSelect.appendChild(sourceOption);

            const targetOption = document.createElement('option');
            targetOption.value = lang.code;
            targetOption.textContent = lang.name;
            targetLanguageSelect.appendChild(targetOption);
        });
    }

    /**
     * Load saved language preferences from localStorage
     */
    function loadSavedLanguages() {
        const savedSource = localStorage.getItem(STORAGE_KEYS.sourceLang);
        const savedTarget = localStorage.getItem(STORAGE_KEYS.targetLang);

        if (savedSource && LANGUAGES.some(l => l.code === savedSource)) {
            sourceLanguageSelect.value = savedSource;
        } else {
            sourceLanguageSelect.value = 'en';
        }

        if (savedTarget && LANGUAGES.some(l => l.code === savedTarget)) {
            targetLanguageSelect.value = savedTarget;
        } else {
            targetLanguageSelect.value = 'es';
        }
    }

    /**
     * Save language preferences to localStorage
     */
    function saveLanguages() {
        localStorage.setItem(STORAGE_KEYS.sourceLang, sourceLanguageSelect.value);
        localStorage.setItem(STORAGE_KEYS.targetLang, targetLanguageSelect.value);
    }

    /**
     * Setup event listeners
     */
    function setupEventListeners() {
        // Translate button
        translateBtn.addEventListener('click', performTranslation);

        // Stop button
        stopBtn.addEventListener('click', stopTranslation);

        // Swap languages button
        swapBtn.addEventListener('click', swapLanguages);

        // Save language preferences on change
        sourceLanguageSelect.addEventListener('change', saveLanguages);
        targetLanguageSelect.addEventListener('change', saveLanguages);

        // Keyboard shortcut: Ctrl+Enter to translate
        sourceTextArea.addEventListener('keydown', (e) => {
            if (e.ctrlKey && e.key === 'Enter') {
                e.preventDefault();
                performTranslation();
            }
        });
    }

    /**
     * Swap source and target languages and text
     */
    function swapLanguages() {
        // Swap language selections
        const tempLang = sourceLanguageSelect.value;
        sourceLanguageSelect.value = targetLanguageSelect.value;
        targetLanguageSelect.value = tempLang;

        // Swap text content
        const tempText = sourceTextArea.value;
        sourceTextArea.value = targetTextArea.value;
        targetTextArea.value = tempText;

        // Save preferences
        saveLanguages();
    }

    /**
     * Perform translation with SSE streaming
     */
    async function performTranslation() {
        const text = sourceTextArea.value.trim();
        if (!text) {
            return;
        }

        const sourceLang = sourceLanguageSelect.value;
        const targetLang = targetLanguageSelect.value;

        if (sourceLang === targetLang) {
            targetTextArea.value = text;
            return;
        }

        // Abort any existing translation
        stopTranslation();

        // Create new abort controller
        abortController = new AbortController();

        // Update UI state
        translateBtn.classList.add('hidden');
        stopBtn.classList.remove('hidden');
        targetTextArea.value = '';
        targetTextArea.classList.add('streaming');

        try {
            const response = await fetch('/api/v1/translate', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({
                    text: text,
                    source_lang: sourceLang,
                    target_lang: targetLang
                }),
                signal: abortController.signal
            });

            if (!response.ok) {
                throw new Error(`HTTP error: ${response.status}`);
            }

            const reader = response.body.getReader();
            const decoder = new TextDecoder();
            let buffer = '';

            while (true) {
                const { done, value } = await reader.read();

                if (done) {
                    break;
                }

                buffer += decoder.decode(value, { stream: true });

                // Process complete SSE messages
                const lines = buffer.split('\n');
                buffer = lines.pop() || '';

                for (const line of lines) {
                    if (line.startsWith('data: ')) {
                        const data = line.slice(6);
                        try {
                            const parsed = JSON.parse(data);

                            if (parsed.type === 'content') {
                                targetTextArea.value += parsed.content;
                            } else if (parsed.type === 'done') {
                                // Translation complete
                            } else if (parsed.type === 'error') {
                                targetTextArea.value = `Error: ${parsed.content}`;
                            }
                        } catch (e) {
                            // Skip invalid JSON
                        }
                    }
                }
            }
        } catch (error) {
            if (error.name === 'AbortError') {
                // Translation was intentionally stopped
            } else {
                targetTextArea.value = `Error: ${error.message}`;
            }
        } finally {
            // Reset UI state
            translateBtn.classList.remove('hidden');
            stopBtn.classList.add('hidden');
            targetTextArea.classList.remove('streaming');
            abortController = null;
        }
    }

    /**
     * Stop the current translation
     */
    function stopTranslation() {
        if (abortController) {
            abortController.abort();
            abortController = null;
        }
    }

    /**
     * Check if translation is in progress
     */
    function isTranslating() {
        return abortController !== null;
    }

    return {
        init,
        performTranslation,
        stopTranslation,
        swapLanguages,
        isTranslating
    };
})();

File: backend/routers/translate.py
"""Translation endpoint with SSE streaming using Tilde."""

import json
from typing import AsyncGenerator

import httpx
from fastapi import APIRouter, Request
from fastapi.responses import StreamingResponse

from backend.auth import get_client_ip, require_user
from backend.config import settings
from backend.models import TranslateRequest
from backend.services.storage import get_admin_settings, get_server, log_activity

router = APIRouter(prefix="/translate", tags=["translate"])

# Language code to name mapping
LANGUAGE_NAMES = {
    "af": "Afrikaans", "sq": "Albanian", "am": "Amharic", "ar": "Arabic",
    "hy": "Armenian", "az": "Azerbaijani", "eu": "Basque", "be": "Belarusian",
    "bn": "Bengali", "bs": "Bosnian", "bg": "Bulgarian", "ca": "Catalan",
    "zh": "Chinese", "hr": "Croatian", "cs": "Czech", "da": "Danish",
    "nl": "Dutch", "en": "English", "et": "Estonian", "fi": "Finnish",
    "fr": "French", "gl": "Galician", "ka": "Georgian", "de": "German",
    "el": "Greek", "gu": "Gujarati", "he": "Hebrew", "hi": "Hindi",
    "hu": "Hungarian", "is": "Icelandic", "id": "Indonesian", "ga": "Irish",
    "it": "Italian", "ja": "Japanese", "kn": "Kannada", "kk": "Kazakh",
    "ko": "Korean", "lv": "Latvian", "lt": "Lithuanian", "mk": "Macedonian",
    "ms": "Malay", "ml": "Malayalam", "mt": "Maltese", "mr": "Marathi",
    "mn": "Mongolian", "ne": "Nepali", "no": "Norwegian", "fa": "Persian",
    "pl": "Polish", "pt": "Portuguese", "pa": "Punjabi", "ro": "Romanian",
    "ru": "Russian", "sr": "Serbian", "si": "Sinhala", "sk": "Slovak",
    "sl": "Slovenian", "es": "Spanish", "sw": "Swahili", "sv": "Swedish",
    "ta": "Tamil", "te": "Telugu", "th": "Thai", "tr": "Turkish",
    "uk": "Ukrainian", "ur": "Urdu", "uz": "Uzbek", "vi": "Vietnamese",
    "cy": "Welsh", "zu": "Zulu"
}


async def stream_translation(
    text: str,
    source_lang: str,
    target_lang: str
) -> AsyncGenerator[str, None]:
    """Stream translation response from configured translation server."""
    # Resolve translation server URL: admin setting > config fallback
    server_url = settings.tilde.url
    try:
        admin_settings = await get_admin_settings()
        if admin_settings.translation_server_id:
            server = await get_server(admin_settings.translation_server_id)
            if server:
                server_url = server.url
    except Exception:
        pass  # Fall back to config

    source_name = LANGUAGE_NAMES.get(source_lang, source_lang)
    target_name = LANGUAGE_NAMES.get(target_lang, target_lang)

    # Build prompt for translation - be specific to avoid extra commentary
    prompt = f"Translate this text from {source_name} to {target_name}. Only provide the translation, nothing else:\n\n{text}"

    # Build request payload in OpenAI chat format
    # Limit tokens to roughly 2x the input to avoid excessive output
    payload = {
        "messages": [{"role": "user", "content": prompt}],
        "stream": True,
        "max_tokens": max(256, len(text) * 3),
        "reasoning_budget": 0,
    }

    try:
        async with httpx.AsyncClient(timeout=None) as client:
            async with client.stream(
                "POST",
                f"{server_url}/v1/chat/completions",
                json=payload,
                headers={"Content-Type": "application/json"}
            ) as response:
                if response.status_code != 200:
                    error_text = await response.aread()
                    yield f"data: {json.dumps({'type': 'error', 'content': f'Translation server error: {error_text.decode()}'})}\n\n"
                    return

                # Buffer to detect special tokens that may be split across chunks
                token_buffer = ""
                stop_streaming = False
                in_think = False  # Track <think>...</think> blocks

                async for line in response.aiter_lines():
                    if not line or stop_streaming:
                        continue

                    if line.startswith("data: "):
                        data = line[6:]

                        if data == "[DONE]":
                            break

                        try:
                            chunk = json.loads(data)
                            if "choices" in chunk and len(chunk["choices"]) > 0:
                                delta = chunk["choices"][0].get("delta", {})
                                content = delta.get("content", "")

                                if content:
                                    # Add to buffer to detect special tokens
                                    token_buffer += content

                                    # Strip <think>...</think> blocks (reasoning models)
                                    while "<think>" in token_buffer:
                                        before = token_buffer[:token_buffer.find("<think>")]
                                        after = token_buffer[token_buffer.find("<think>") + 7:]
                                        if "</think>" in after:
                                            token_buffer = before + after[after.find("</think>") + 8:]
                                        else:
                                            # Still inside think block - send what's before it
                                            if before:
                                                yield f"data: {json.dumps({'type': 'content', 'content': before})}\n\n"
                                            token_buffer = after
                                            in_think = True
                                            break

                                    # If inside a think block, look for closing tag
                                    if in_think:
                                        if "</think>" in token_buffer:
                                            token_buffer = token_buffer[token_buffer.find("</think>") + 8:]
                                            in_think = False
                                        else:
                                            # Discard buffered think content, keep tail for partial tag
                                            token_buffer = token_buffer[-8:] if len(token_buffer) > 8 else token_buffer
                                            continue

                                    # Strip stray </think> tags
                                    token_buffer = token_buffer.replace("</think>", "")

                                    if in_think:
                                        continue

                                    # Check if buffer contains start of special token
                                    if "<|" in token_buffer:
                                        # Output everything before the special token
                                        idx = token_buffer.find("<|")
                                        if idx > 0:
                                            yield f"data: {json.dumps({'type': 'content', 'content': token_buffer[:idx]})}\n\n"
                                        stop_streaming = True
                                        continue

                                    # If buffer is getting long and no special token, flush it
                                    if len(token_buffer) > 10:
                                        # Keep last few chars in case token spans chunks
                                        to_send = token_buffer[:-5]
                                        token_buffer = token_buffer[-5:]
                                        if to_send:
                                            yield f"data: {json.dumps({'type': 'content', 'content': to_send})}\n\n"
                        except json.JSONDecodeError:
                            continue

                # Flush remaining buffer if no special token detected
                if token_buffer and not stop_streaming and not in_think and "<|" not in token_buffer:
                    yield f"data: {json.dumps({'type': 'content', 'content': token_buffer})}\n\n"

        yield f"data: {json.dumps({'type': 'done'})}\n\n"

    except httpx.ConnectError:
        yield f"data: {json.dumps({'type': 'error', 'content': 'Cannot connect to translation server. Is Tilde running?'})}\n\n"
    except Exception as e:
        yield f"data: {json.dumps({'type': 'error', 'content': str(e)})}\n\n"


@router.post("")
async def translate_text(data: TranslateRequest, request: Request):
    """Translate text and stream the response."""
    await require_user(request)
    ip = get_client_ip(request)
    source_name = LANGUAGE_NAMES.get(data.source_lang, data.source_lang)
    target_name = LANGUAGE_NAMES.get(data.target_lang, data.target_lang)
    await log_activity(ip, "translate", f"Translated from {source_name} to {target_name}")
    return StreamingResponse(
        stream_translation(data.text, data.source_lang, data.target_lang),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no",
        }
    )
